lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1417 @@
+    $ret = $nicByMac + $nicByName
+    if ($ret.Length -eq 0 -and $MustFindAdapter){
+        Throw "Could not find network adapters"
+    }
+    return $ret
+}
+
+
+function Get-DataPortFromDataNetwork {
+    $dataNetwork = Get-JujuCharmConfig -Scope "os-data-network"
+    if (!$dataNetwork) {
+        Write-JujuInfo "os-data-network is not defined"
+        return $false
+    }
+
+    $local_ip = Get-CharmState -Namespace "hvcomputesrc" -Key "local_ip"
+    $ifIndex = Get-CharmState -Namespace "hvcomputesrc" -Key "dataNetworkIfindex"
+
+    if($local_ip -and $ifIndex){
+        if((Confirm-LocalIP -IPaddress $ifIndex -ifIndex $ifIndex)){
+            return Get-NetAdapter -ifindex $ifIndex
+        }
+    }
+
+    # If there is any network interface configured to use DHCP and did not get an IP address
+    # we manually renew its lease and try to get an IP address before searching for the data network
+    $interfaces = Get-CimInstance -Class win32_networkadapterconfiguration | Where-Object { 
+        $_.IPEnabled -eq $true -and $_.DHCPEnabled -eq $true -and $_.DHCPServer -eq "255.255.255.255"
+    }
+    if($interfaces){
+        $interfaces.InterfaceIndex | Invoke-DHCPRenew -ErrorAction SilentlyContinue
+    }
+    $netDetails = $dataNetwork.Split("/")
+    $decimalMask = ConvertTo-Mask $netDetails[1]
+
+    $configuredAddresses = Get-NetIPAddress -AddressFamily IPv4
+    foreach ($i in $configuredAddresses) {
+        Write-JujuInfo ("Checking {0} on interface {1}" -f @($i.IPAddress, $i.InterfaceAlias))
+        if ($i.PrefixLength -ne $netDetails[1]){
+            continue
+        }
+        $network = Get-NetworkAddress $i.IPv4Address $decimalMask
+        Write-JujuInfo ("Network address for {0} is {1}" -f @($i.IPAddress, $network))
+        if ($network -eq $netDetails[0]){
+            Set-CharmState -Namespace "hvcomputesrc" -Key "local_ip_mask" -Value $i.PrefixLength
+            Set-CharmState -Namespace "hvcomputesrc" -Key "local_ip" -Value $i.IPAddress
+            Set-CharmState -Namespace "hvcomputesrc" -Key "dataNetworkIfindex" -Value $i.IfIndex
+            return Get-NetAdapter -ifindex $i.IfIndex
+        }
+    }
+    return $false
+}
+
+
+function Get-RealInterface {
+    [CmdletBinding()]
+    Param(
+        [Parameter(Mandatory=$true, ValueFromPipeline=$true)]
+        [Microsoft.Management.Infrastructure.CimInstance]$interface
+    )
+    PROCESS {
+        if($interface.DriverFileName -ne "vmswitch.sys") {
+            return $interface
+        }
+        $realInterface = Get-NetAdapter | Where-Object {
+            $_.MacAddress -eq $interface.MacAddress -and $_.ifIndex -ne $interface.ifIndex
+        }
+
+        if(!$realInterface){
+            Throw "Failed to find interface attached to VMSwitch"
+        }
+        return $realInterface[0]
+    }
+}
+
+
+function Get-FallbackNetadapter {
+    $name = Get-MainNetadapter
+    $net = Get-NetAdapter -Name $name
+    return $net
+}
+
+
+function Get-OVSDataPort {
```

Comment:
Something similar (import it if it's what you need) -> [here](https://github.com/cloudbase/nova-hyperv-charm/blob/master/lib/Modules/OVSCharmUtils/OVSCharmUtils.psm1#L129)